src/api/mod.rs

```diff
@@ -403,6 +440,117 @@ impl<'a> ValueListBuilder<'a> {
     }
 }
 
+fn to_meta_data<'a, T>(meta_hm: T) -> Result<*mut meta_data_t, SubmitError> 
+    where T : IntoIterator<Item=(&'a String, &'a MetaValue)>
+{
+    let meta = unsafe { meta_data_create() };
+    for (key, value) in meta_hm.into_iter() {
+        let c_key = CString::new(key.as_str()).map_err(|e| {
+            SubmitError::Field(
+                "meta key",
+                ArrayError::NullPresent(e.nul_position(), key.to_string()),
+            )
+        })?;
+        match value {
+            MetaValue::String(str) => {
+                let c_value = CString::new(str.as_str()).map_err(|e| {
+                    SubmitError::Field(
+                        "meta value",
+                        ArrayError::NullPresent(e.nul_position(), str.to_string()),
+                    )
+                })?;
+                unsafe {
+                    meta_data_add_string(meta, c_key.as_ptr(), c_value.as_ptr());
+                }
+            }
+            MetaValue::SignedInt(i) => unsafe {
+                meta_data_add_signed_int(meta, c_key.as_ptr(), *i);
+            },
+            MetaValue::UnsignedInt(u) => unsafe {
+                meta_data_add_unsigned_int(meta, c_key.as_ptr(), *u);
+            },
+            MetaValue::Double(d) => unsafe {
+                meta_data_add_double(meta, c_key.as_ptr(), *d);
+            },
+            MetaValue::Boolean(b) => unsafe {
+                meta_data_add_boolean(meta, c_key.as_ptr(), *b);
+            },
+        }
+    }
+    Ok(meta)
+}
+
+fn from_meta_data(
+    p: &str,
+    meta: *mut meta_data_t,
+) -> Result<Option<HashMap<String, MetaValue>>, ReceiveError> {
+    if meta.is_null() {
+        return Ok(None);
+    }
+
+    let mut c_toc: *mut *mut c_char = ptr::null_mut();
+    let count_or_err = unsafe { meta_data_toc(meta, &mut c_toc as *mut *mut *mut c_char) };
+    if count_or_err < 0 {
+        return Err(ReceiveError::Metadata(p.to_string(), "toc".to_string(), "invalid parameters to meta_data_toc"));
+    }
+    let count = count_or_err as usize;
+    if count == 0 {
+        return Ok(None);
+    }
+    let mut meta_hm = HashMap::new();
+    let toc = unsafe { slice::from_raw_parts(c_toc, count) };
+    // read meta and convert to hashmap entry
+    for c_key_ptr in toc {
+        let (c_key, key, value_type) = unsafe {
+            let c_key : &CStr = CStr::from_ptr(*c_key_ptr);
+            let key : String = c_key.to_str().map_err(|e| ReceiveError::Utf8(p.to_string(), "metadata key", e))?.to_string();
```

Comment:
I think this is an instance where key can be a `&str` instead of a `String` as other fields in a `ValueList` (like `host`) use string slices.

Reply:
done

Reply:
here i don't see how to do it.
the key data is owned by `toc`, if i use a reference to it i cannot drop `toc` anymore.
So i should store the `toc` ptr in `ValueList`, and call libc::free on drop ?

It seems safer to copy the data to a string and immediatly drop `toc` data at the end of the function.
Sorry I am not accustomed enough to rust and ffi to do this i think.

Reply:
while thinking about this, it occured to me that if a ReceiveError is returned, the `toc` leaks.

Reply:
I agree with your thoughts. Just the fixup to ensure that toc doesn't leak sounds good.

Reply:
on it.
I'll do the same in to_meta_data where the meta_data_t leaks if an error occurs

Reply:
done.